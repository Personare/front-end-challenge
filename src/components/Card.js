import React, { PropTypes } from 'react';
import TweenLite from 'gsap';

export class Card extends React.Component {
    componentWillReceiveProps(nextProps) {
        if (nextProps.selected !== this.props.selected) {
            TweenLite.to(this.refs.cardFaces, 1, { rotationY: !nextProps.selected ? -180 : 0 });
        }
    }

    render() {
        return (
            <article className="card" onClick={this.props.onClick}>
                <div className="card-faces" ref="cardFaces">
                    <div className="front">
                        <img src={this.props.frontImg} alt={this.props.name} />
                    </div>
                    <div className="back">
                        <img src={this.props.backImg} alt={this.props.name} />
                    </div>
                </div>
            </article>
        );
    }
}

Card.propTypes = {
    name: PropTypes.string,
    frontImg: PropTypes.string,
    backImg: PropTypes.string,
    selected: PropTypes.bool,
    onClick: PropTypes.func,
};
