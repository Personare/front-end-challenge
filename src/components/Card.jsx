import React from "react";

const Card = ({ name = "CartjÃªnha", image = "", ...rest }) => {
  return (
    <div className="card">
      <img src={image} alt={name} />
    </div>
  );
};

export default Card;
