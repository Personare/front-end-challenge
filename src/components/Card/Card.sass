.CardContainer
  position: relative
  height: 210px
  cursor: pointer
  margin: 10px;
  .CardContent
    transform-style: preserve-3d

  .CardImage, .BackImage
    position: absolute
    width: 100px
    height: 210px
    overflow: hidden
    border-radius: 5px
    backface-visibility: hidden
    > img
      width: 100px
      height: 210px

  .BackImage
    transform: rotateY(-180deg)
