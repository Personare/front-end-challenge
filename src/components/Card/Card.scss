.Card {
    position: relative;
    perspective: 150rem;
    width: 200px;
    height: 200px;

    &__side {
        position: absolute;
        top: 0;
        left: 0;
        width: 100%;
        height: 50px;
        backface-visibility: hidden;
        transition: all .8s ease;

        &.--front,
        &.--back {
            svg {
                position: absolute;
                top: 0;
                right: 0;
                bottom: 0;
                left: 0;
                width: 1em;
                height: 1em;
                font-size: 5.4rem;
            }
        }

        &.--front {
            background-color: red;

            svg {
                fill: black;
            }
        }

        &.--back {
            background-color: blue;
            transform: rotateY(180deg);

            svg {
                fill: red;
            }
        }
    }

    &.--closed &__side.\--front {
        transform: rotateY(-180deg);
    }

    &.--closed &__side.\--back {
        transform: rotateY(0);
    }
}

