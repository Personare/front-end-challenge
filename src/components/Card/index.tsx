import { useContext } from "react";
import ReactCardFlip from "react-card-flip";
import { ICard } from "../../config/interfaces";
import GlobalContext from "../../global/GlobalContext";

interface ICardItem {
  card: ICard;
  onClick?: (name: string) => void;
}

const Card: React.FC<ICardItem> = ({ card, onClick }) => {
  const { name, image, show } = card;

  const {
    state: { data },
  }: any = useContext(GlobalContext);

  const isFlipped = !show;

  console.log(onClick);

  return (
    <div data-cy="card" onClick={() => onClick && onClick(name)}>
      <ReactCardFlip isFlipped={isFlipped} flipDirection="horizontal">
        <img
          alt={name}
          src={`${data.imagesUrl}${image}`}
          data-cy={!isFlipped && "card-up"}
        />

        <img
          alt={name}
          src={data.imageBackCard}
          data-cy={isFlipped && "card-down"}
        />
      </ReactCardFlip>
    </div>
  );
};

export default Card;
