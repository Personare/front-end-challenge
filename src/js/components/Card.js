import React, { Component } from 'react';

class Card extends Component {
    constructor(props) {
        super(props);

        this.updateDimensions = this.updateDimensions.bind(this);
    }

    componentDidMount() {
        window.addEventListener("resize", this.updateDimensions);

        this.updateDimensions();
    }

    componentWillUnmount() {
        window.removeEventListener("resize", this.updateDimensions);
    }

    updateDimensions() {
        const { cardX, cardY, cardIndex } = this.props;
        const element = this.refs[cardIndex];

        let totalMargens = cardX * 20;
        let leftPosition = cardX * element.offsetWidth + totalMargens;

        let totalTopMargins = cardY * 20;
        let topPosition = cardY * element.offsetHeight + totalTopMargins;

        let parent = element.parentNode;
        let parentHeight = (element.offsetHeight + 20) * (cardY + 1) - 20;

        parent.style.height = `${parentHeight}px`;
        element.style.left = `${leftPosition}px`;
        element.style.top = `${topPosition}px`;
    }

    render() {
        return (
            <div className='Card'
                 style={{ left: this.props.cardX, top: this.props.cardY }}
                 ref={this.props.cardIndex}
            >
                <div className='Card__wrapper'>
                    <div className='Card__side is-front'>
                        <img className='Card__image' src={this.props.imagesUrl} />
                    </div>
                    <div className='Card__side is-back h-hide'>
                        <img className='Card__image' src={this.props.imageBackCard} />
                    </div>
                </div>
            </div>
        );
    }
};

export default Card;
