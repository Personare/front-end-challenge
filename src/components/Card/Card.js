import React, { Component } from 'react';
import TweenMax from 'gsap';

import './Card.sass';

export class Card extends Component {
  constructor(props) {
    super(props);
    this.cardContentRef = React.createRef();
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.selected !== this.props.selected) {
      TweenMax.to(this.cardContentRef.current, 1, { rotationY: !nextProps.selected ? -180 : 0 });
    }
  }

  render() {
    return (
      <article className="CardContainer" onClick={this.props.click}>
        <div className="CardContent" ref={this.cardContentRef}>
          <div className="CardImage">
            <img src={this.props.card} alt={this.props.name} />
          </div>
          <div className="BackImage">
            <img src={this.props.imageBack} alt={this.props.name} />
          </div>
        </div>
      </article>
    );
  }
}

export default Card;
