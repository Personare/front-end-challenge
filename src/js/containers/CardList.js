import React, { Component } from 'react';
import { connect } from 'react-redux';
import { changeGridCols, flipCardThenOpenModal } from '../actions/game';
import { getBreakpoint } from '../helpers/utils';
import Card from '../components/Card';

class CardList extends Component {
    constructor(props) {
        super(props);
        this.setGridColumn = this.setGridColumn.bind(this);
    }

    componentDidMount() {
        window.addEventListener("resize", this.setGridColumn);
    }

    componentWillUnmount() {
        window.removeEventListener("resize", this.setGridColumn);
    }

    componentWillMount() {
        this.setGridColumn();
    }

    setGridColumn() {
        var currentBreakpoint = getBreakpoint();
        let totalCols = 6; // desktop default

        if (currentBreakpoint === 'tablet') {
            totalCols = 6;
        } else if (currentBreakpoint === 'phone') {
            totalCols = 3;
        } else if (currentBreakpoint === 'smallPhone') {
            totalCols = 2;
        }

        if (totalCols === this.props.gridCollumn) {
            return false;
        }

        this.props.setGridCols(totalCols);
    }


    render() {
        if(this.props.isLoading) {
            return (<span>loading...</span>);
        }

        const { imageBackCard, imagesUrl, cards } = this.props.data;

        const { animationMode, gridCollumn } = this.props;
        const totalCards = cards.length;

        const cardsItems = cards.map((card, index) => {
            const subValue = card.position % gridCollumn === 0 ? 0 : 1;

            const cardX = (card.position % gridCollumn);
            const cardY = (Math.ceil(card.position / gridCollumn) - subValue);

            let isShuffling = animationMode === 'joining';

            return (
                <Card
                    key={index}
                    name={card.name}
                    imageBackCard={imageBackCard}
                    imagesUrl={imagesUrl + card.image}
                    cardIndex={index}
                    cardX={cardX}
                    cardY={cardY}
                    flipped={card.flipped}
                    shuffling={isShuffling}
                    totalCards={totalCards}
                    gridCollumn={gridCollumn}
                    onClick={this.props.flipCardThenOpenModal}
                />
            );
        });

        return (
            <div className='CardList'>
                {cardsItems}
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        data: state.data,
        isLoading: state.gameState.isLoading,
        animationMode: state.gameState.animationMode,
        gridCollumn: state.gameState.gridCollumn
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        flipCardThenOpenModal: (cardID) => {
            dispatch(flipCardThenOpenModal(cardID));
        },
        setGridCols: (totalCols) => {
            dispatch(changeGridCols(totalCols))
        },
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(CardList);
