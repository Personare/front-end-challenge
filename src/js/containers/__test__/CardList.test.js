import React from 'react';
import Enzyme, { shallow, render, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Unwrapped as UnwrappedCardList } from '../CardList';
import Card from '../../components/Card';

// setup enzyme
Enzyme.configure({ adapter: new Adapter() });

const getMockCards = () => {
    return {
        "imagesUrl": "https://dkw5ssdvaqf8l.cloudfront.net/static/psr/br/framework/yii/images/content/pt-br/product/tarot/marselha/162x341/",
        "imageBackCard": "https://dkw5ssdvaqf8l.cloudfront.net/static/psr/br/framework/yii/images/content/pt-br/product/tarot/marselha/back-blue-card.png",
        "cards": [
            {
              "name": "O MAGO",
              "image": "arcano1.jpg",
              "flipped": false,
              "position": 0
            },
            {
              "name": "A SACERDOTISA",
              "image": "arcano2.jpg",
              "flipped": false,
              "position": 1
            },
            {
              "name": "A IMPERATRIZ",
              "image": "arcano3.jpg",
              "flipped": false,
              "position": 2
            }
        ]
    };
};

function getMockProps() {
    return {
        data: getMockCards(),
        isLoading: false,
        animationMode: '',
        gridCollumn: 6
    };
}

describe("CardList component", () => {

    it('renders correctly', () => {
        const props = getMockProps();
        const cardList = shallow(<UnwrappedCardList {...props}/>);

        expect(cardList).toMatchSnapshot();
    });

    it('renders the right amount of cards', () => {
        const props = getMockProps();
        const cardList = shallow(<UnwrappedCardList {...props} />);
        const cardsCount = props.data.cards.length;

        expect(cardList.find(Card).length).toBe(cardsCount);
    });

});
