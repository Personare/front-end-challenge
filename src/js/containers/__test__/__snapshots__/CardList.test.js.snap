// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardList component renders correctly 1`] = `
<div
  className="CardList"
  style={
    Object {
      "height": "5000px",
    }
  }
>
  <Card
    cardIndex={0}
    cardX={0}
    cardY={0}
    flipped={false}
    gridCollumn={6}
    imageBackCard="https://dkw5ssdvaqf8l.cloudfront.net/static/psr/br/framework/yii/images/content/pt-br/product/tarot/marselha/back-blue-card.png"
    imagesUrl="https://dkw5ssdvaqf8l.cloudfront.net/static/psr/br/framework/yii/images/content/pt-br/product/tarot/marselha/162x341/arcano1.jpg"
    key="0"
    name="O MAGO"
    shuffling={false}
    totalCards={3}
  />
  <Card
    cardIndex={1}
    cardX={1}
    cardY={0}
    flipped={false}
    gridCollumn={6}
    imageBackCard="https://dkw5ssdvaqf8l.cloudfront.net/static/psr/br/framework/yii/images/content/pt-br/product/tarot/marselha/back-blue-card.png"
    imagesUrl="https://dkw5ssdvaqf8l.cloudfront.net/static/psr/br/framework/yii/images/content/pt-br/product/tarot/marselha/162x341/arcano2.jpg"
    key="1"
    name="A SACERDOTISA"
    shuffling={false}
    totalCards={3}
  />
  <Card
    cardIndex={2}
    cardX={2}
    cardY={0}
    flipped={false}
    gridCollumn={6}
    imageBackCard="https://dkw5ssdvaqf8l.cloudfront.net/static/psr/br/framework/yii/images/content/pt-br/product/tarot/marselha/back-blue-card.png"
    imagesUrl="https://dkw5ssdvaqf8l.cloudfront.net/static/psr/br/framework/yii/images/content/pt-br/product/tarot/marselha/162x341/arcano3.jpg"
    key="2"
    name="A IMPERATRIZ"
    shuffling={false}
    totalCards={3}
  />
</div>
`;
